From 4bf3613e3b1a7bf4996b140daee59d81156fb373 Mon Sep 17 00:00:00 2001
From: Xie XiuQi <xiexiuqi@huawei.com>
Date: Fri, 26 Feb 2021 20:21:29 +0800
Subject: [PATCH] arm64/mpam: only add new domain node to domain list
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: 4bf3613e3b1a7bf4996b140daee59d81156fb373


hulk inclusion
category: bugfix
bugzilla: 48265
CVE: NA

--------------------------------

Only add a new domain to domain list, don't care about exist domain
node.

Signed-off-by: Xie XiuQi <xiexiuqi@huawei.com>
Reviewed-by: Hanjun Guo <guohanjun@huawei.com>
Signed-off-by: Yang Yingliang <yangyingliang@huawei.com>
Signed-off-by: Wang ShaoBo <bobo.shaobowang@huawei.com>
Reviewed-by: Cheng Jian <cj.chengjian@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 arch/arm64/kernel/mpam.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/arch/arm64/kernel/mpam.c b/arch/arm64/kernel/mpam.c
index a83b43c9dcaa..c38c13bde995 100644
--- a/arch/arm64/kernel/mpam.c
+++ b/arch/arm64/kernel/mpam.c
@@ -1208,6 +1208,8 @@ static void mpam_domains_init(struct resctrl_resource *r)
 
 		if (!d)
 			d = kzalloc(sizeof(*d), GFP_KERNEL);
+		else
+			continue;
 
 		if (!d) {
 			mpam_domains_destroy(r);
@@ -1263,7 +1265,8 @@ static void mpam_domains_init(struct resctrl_resource *r)
 			return;
 		}
 
-		list_add_tail(&d->list, add_pos);
+		if (add_pos)
+			list_add_tail(&d->list, add_pos);
 
 		id++;
 	}
-- 
2.26.2

